File: Mar 27 1337. The K Weakest Rows in a Matrix (Easy)/Solution.py
# Linear Search, Sort O(M*N + MlogM) ) | O(M)
class Solution:
    def kWeakestRows(self, mat: List[List[int]], k: int) -> List[int]:
        countRow = []
        for rowIdx, row in enumerate(mat):
            count = sum(row)
            countRow.append([rowIdx, count])
        countRow.sort(key = lambda x: x[1])
        res = [0] * k
        for i in range(k):
            res[i] = countRow[i][0]    
        return res

# Binary Search, Sort O(MlogN + MlogM) ) | O(M)
class Solution:
    def kWeakestRows(self, mat: List[List[int]], k: int) -> List[int]:
        countRow = []
        # Binary Search for each row
        for rowIdx, row in enumerate(mat):
            count = self.binarySearch(row)
            countRow.append([rowIdx, count])
        # Sort countRow
        countRow.sort(key = lambda x: x[1])
        res = [0] * k
        for i in range(k):
            res[i] = countRow[i][0]    
        return res

    # Binary Search
    def binarySearch(self, arr):
        left, right = 0, len(arr) - 1
        while left <= right:
            mid = (left + right) // 2
            if arr[mid] == 1:
                left = mid + 1
            else:
                right = mid - 1
        return left

# Binary Search, Heap O(MlogN + MlogM) ) | O(M)
import heapq
class Solution:
    def kWeakestRows(self, mat: List[List[int]], k: int) -> List[int]:
        queue = []
        # Binary Search for each row
        for rowIdx, row in enumerate(mat):
            count = self.binarySearch(row)
            entry = [count , rowIdx]
            heapq.heappush(queue, entry)
                
        res = [0] * k
        for i in range(k):
            count, idx = heapq.heappop(queue)
            res[i] = idx
        return res

    # Binary Search
    def binarySearch(self, arr):
        left, right = 0, len(arr) - 1
        while left <= right:
            mid = (left + right) // 2
            if arr[mid] == 1:
                left = mid + 1
            else:
                right = mid - 1
        return left
        
# MaxHeap and Reverse Heap O(MlogN + Mlogk) ) | O(M)
import heapq

class Solution:
    def kWeakestRows(self, mat: List[List[int]], k: int) -> List[int]:
        queue = []
        # Binary Search for each row
        for rowIdx, row in enumerate(mat):
            count = self.binarySearch(row)
            entry = [-count , -rowIdx]
            if len(queue) < k or entry > queue[0]:
                heapq.heappush(queue, entry)
            if len(queue) > k:
                heapq.heappop(queue)
                
        res = [0] * k
        for i in range(k):
            count, idx = heapq.heappop(queue)
            res[i] = -idx
        return res[::-1]

    # Binary Search
    def binarySearch(self, arr):
        left, right = 0, len(arr) - 1
        while left <= right:
            mid = (left + right) // 2
            if arr[mid] == 1:
                left = mid + 1
            else:
                right = mid - 1
        return left

# Traverse Col by cols O(M * N) | O(k)
class Solution:
    def kWeakestRows(self, mat: List[List[int]], k: int) -> List[int]:
        res = []
        for col in range(len(mat[0])):
            for row in range(len(mat)):
                if mat[row][col] == 0:
                    if col == 0 or mat[row][col - 1] != 0:
                        res.append(row)
                if len(res) == k:
                    return res
                
        for row in range(len(mat)):
            if mat[row][-1] == 1:
                res.append(row)
            if len(res) == k:
                return res                 

File: Mar 27 1337. The K Weakest Rows in a Matrix (Easy)/Solution.go
// Linear Search, Sort O(M*N + MlogM) ) | O(M) Approach 1 ---------------------------------------------
func kWeakestRows(mat [][]int, k int) []int {
    countRow := make([][]int, 0)
    for rowIdx, row := range(mat){
        count := 0
        for _, num := range(row){
            count += num
        }
        countRow = append(countRow, []int{rowIdx, count})
    }
    
    sort.Slice(countRow, func(i, j int)bool{
        if countRow[i][1] == countRow[j][1]{
            return countRow[i][0] < countRow[j][0]
        }
        return countRow[i][1] < countRow[j][1]
    })
    res := make([]int, k)
    for i:=0; i < k; i++{
        res[i] = countRow[i][0]
    }
    return res
}

// Binary Search, Sort O(MlogN + MlogM) ) | O(M) Approach 2 ---------------------------------------------
func kWeakestRows(mat [][]int, k int) []int {
    countRow := make([][]int, 0)
    for rowIdx, row := range(mat){
        count := binarySearch(row)
        countRow = append(countRow, []int{rowIdx, count})
    }
    
    sort.Slice(countRow, func(i, j int)bool{
        if countRow[i][1] == countRow[j][1]{
            return countRow[i][0] < countRow[j][0]
        }
        return countRow[i][1] < countRow[j][1]
    })
    res := make([]int, k)
    for i:=0; i < k; i++{
        res[i] = countRow[i][0]
    }
    return res
}

func binarySearch(arr []int) int{
    left, right := 0, len(arr) - 1
    for left <= right{
        mid := (left + right) / 2
        if arr[mid] == 1{
            left = mid + 1
        } else{
            right = mid - 1
        }
    }
    return left
}

// Binary Search, Heap O(MlogN + MlogM) ) | O(M) Approach 3 ---------------------------------------------
//---------Build a Heap---------//
type Item struct {
    idx  int
    count int
}

type ItemHeap []Item

func (h ItemHeap) Len() int { return len(h) }

func (h ItemHeap) Less(i, j int) bool {
    if h[i].count != h[j].count {
        return h[i].count < h[j].count
    } else {
        return h[i].idx < h[j].idx
    }
}

func (h ItemHeap) Swap(i, j int) { h[i], h[j] = h[j], h[i] }

func (h *ItemHeap) Push(val interface{}) {
    *h = append(*h, val.(Item))
}

func (h *ItemHeap) Pop() interface{} {
    old := *h
    n := len(old)
    x := old[n-1]
    *h = old[0 : n-1]
    return x
}
//---------- Solution -----------//
func kWeakestRows(mat [][]int, k int) []int {
    queue := &ItemHeap{}
    for rowIdx, row := range(mat){
        count := binarySearch(row)
        heap.Push(queue, Item{
            idx: rowIdx,
            count: count,
        })
    }
    res := make([]int , k)
    for i :=0; i < k; i++{
        item := heap.Pop(queue).(Item)
        res[i] = item.idx
    }
    return res
}

func binarySearch(arr []int) int{
    left, right := 0, len(arr) - 1
    for left <= right{
        mid := (left + right) / 2
        if arr[mid] == 1{
            left = mid + 1
        } else{
            right = mid - 1
        }
    }
    return left
}

// MaxHeap and Reverse Heap O(MlogN + Mlogk) ) | O(M) Approach 4 ---------------------------------------------

//---------Build a maxHeap--------------
type Item struct {
    idx  int
    count int
}

type ItemHeap []Item

func (h ItemHeap) Len() int { return len(h) }
// Change Less function to Max heap
func (h ItemHeap) Less(i, j int) bool {
    if h[i].count != h[j].count {
        return h[i].count >= h[j].count
    } else {
        return h[i].idx >= h[j].idx
    }
}

func (h ItemHeap) Swap(i, j int) { h[i], h[j] = h[j], h[i] }

func (h *ItemHeap) Push(val interface{}) {
    *h = append(*h, val.(Item))
}

func (h *ItemHeap) Pop() interface{} {
    old := *h
    n := len(old)
    x := old[n-1]
    *h = old[0 : n-1]
    return x
}
//---------- Solution -----------//
func kWeakestRows(mat [][]int, k int) []int {
    queue := &ItemHeap{}
    for rowIdx, row := range(mat){
        count := binarySearch(row)
        if queue.Len() < k || count < (*queue)[0].count{
            heap.Push(queue, Item{
            idx: rowIdx,
            count: count,
            })
        }
        if queue.Len() > k{
            heap.Pop(queue)
        }

    }
    res := make([]int , k)
    for i :=0; i < k; i++{
        item := heap.Pop(queue).(Item)
        res[k - i - 1] = item.idx
    }
    return res
}

func binarySearch(arr []int) int{
    left, right := 0, len(arr) - 1
    for left <= right{
        mid := (left + right) / 2
        if arr[mid] == 1{
            left = mid + 1
        } else{
            right = mid - 1
        }
    }
    return left
}

// Traverse Col by cols O(M * N) | O(k) Approach 5 ---------------------------------------------
func kWeakestRows(mat [][]int, k int) []int {
    res := make([]int, 0)
    for col := 0; col < len(mat[0]); col++{
        for row := 0; row < len(mat); row++{
            if mat[row][col] == 0{
                if col == 0 || mat[row][col - 1] == 1{
                    res = append(res, row)
                }
                if len(res) == k{
                    return res
                }
            }
        }
    }
    
    for row := 0; row < len(mat); row++{
        if mat[row][len(mat[row]) - 1] == 1{
            res = append(res, row)
        }
        if len(res) == k{
            break
        }
    }
    return res
}


